File: src/infrastructure/firestore/FirebaseAttendanceRepository.ts
import { FirebaseError } from 'firebase/app'
import {
  collection,
  doc,
  getDoc,
  getDocs,
  query,
  setDoc,
  where,
  Timestamp,
} from 'firebase/firestore'
import type { Attendance } from '@/domain/entities/Attendance'
import {
  attendanceDocId,
  buildOfficeQrPayload,
  isAttendanceOrigin,
} from '@/domain/entities/Attendance'
import type { AttendanceOfficeQr } from '@/domain/entities/AttendanceOfficeQr'
import type { AttendanceSettings } from '@/domain/entities/AttendanceSettings'
import {
  normalizeAttendanceSettings,
  normalizeOfficeRadiusMeters,
} from '@/domain/entities/AttendanceSettings'
import type {
  AttendanceRepository,
  CreateAttendanceInput,
  SaveAttendanceOfficeQrInput,
  SaveAttendanceSettingsInput,
} from '@/domain/repositories/AttendanceRepository'
import { UnauthorizedError, ValidationError } from '@/domain/errors/DomainError'
import { firestoreDb } from '@/infrastructure/firebase/firebaseApp'

interface AttendanceDoc {
  userId: string
  userName: string
  dateKey: string
  origin: string
  areaId: string
  areaName: string
  latitude: number
  longitude: number
  locationAccuracy?: number
  distanceToOfficeMeters?: number
  officeValidated: boolean
  officeQrToken?: string
  permissionNote?: string
  markedById?: string
  markedByName?: string
  environmentPhotoUrl?: string
  environmentPhotoPath?: string
  createdAt: Timestamp
}

interface SettingsDoc {
  officeName: string
  officeLatitude: number
  officeLongitude: number
  officeRadiusMeters: number
  officePoints?: Array<{
    id: string
    name: string
    latitude: number
    longitude: number
    radiusMeters: number
  }>
  updatedAt: Timestamp
  updatedById: string
  updatedByName: string
}

interface OfficeQrDoc {
  dateKey: string
  token: string
  validFrom: Timestamp
  validUntil: Timestamp
  createdAt: Timestamp
  createdById: string
}

function mapFirebaseWriteError(error: unknown, fallback: string): never {
  if (error instanceof ValidationError || error instanceof UnauthorizedError) {
    throw error
  }
  if (error instanceof FirebaseError && error.code === 'permission-denied') {
    throw new UnauthorizedError(
      'No se pudo registrar. Revisa el GPS, el radio de oficina o si ya hay una marca ese día.',
    )
  }
  throw new ValidationError(fallback)
}

function mapAttendance(id: string, data: AttendanceDoc): Attendance | null {
  if (!isAttendanceOrigin(data.origin)) return null
  return {
    id,
    userId: data.userId,
    userName: data.userName,
    dateKey: data.dateKey,
    origin: data.origin,
    areaId: data.areaId ?? '',
    areaName: data.areaName ?? '',
    latitude: data.latitude,
    longitude: data.longitude,
    accuracyMeters:
      typeof data.locationAccuracy === 'number'
        ? data.locationAccuracy
        : undefined,
    distanceToOfficeMeters:
      typeof data.distanceToOfficeMeters === 'number'
        ? data.distanceToOfficeMeters
        : undefined,
    officeValidated: data.officeValidated === true,
    permissionNote: data.permissionNote?.trim() || undefined,
    markedById: data.markedById || undefined,
    markedByName: data.markedByName || undefined,
    environmentPhotoUrl: data.environmentPhotoUrl || undefined,
    environmentPhotoPath: data.environmentPhotoPath || undefined,
    createdAt: data.createdAt?.toDate?.() ?? new Date(),
  }
}

function mapOfficeQr(data: OfficeQrDoc): AttendanceOfficeQr {
  return {
    dateKey: data.dateKey,
    token: data.token,
    payload: buildOfficeQrPayload(data.dateKey, data.token),
    validFrom: data.validFrom.toDate(),
    validUntil: data.validUntil.toDate(),
    createdAt: data.createdAt.toDate(),
    createdById: data.createdById,
  }
}

export class FirebaseAttendanceRepository implements AttendanceRepository {
  private readonly collectionRef = collection(firestoreDb, 'attendances')
  private readonly settingsRef = doc(firestoreDb, 'settings', 'attendance')
  private readonly officeQrCollection = collection(
    firestoreDb,
    'attendanceOfficeTokens',
  )

  async getSettings(): Promise<AttendanceSettings | null> {
    const snapshot = await getDoc(this.settingsRef)
    if (!snapshot.exists()) return null
    const data = snapshot.data() as SettingsDoc
    return normalizeAttendanceSettings({
      officePoints: data.officePoints,
      officeName: data.officeName,
      officeLatitude: data.officeLatitude,
      officeLongitude: data.officeLongitude,
      officeRadiusMeters: normalizeOfficeRadiusMeters(data.officeRadiusMeters),
      updatedAt: data.updatedAt?.toDate?.() ?? new Date(),
      updatedById: data.updatedById ?? '',
      updatedByName: data.updatedByName ?? '',
    })
  }

  async saveSettings(
    input: SaveAttendanceSettingsInput,
  ): Promise<AttendanceSettings> {
    const now = Timestamp.now()
    const normalized = normalizeAttendanceSettings({
      officePoints: input.officePoints,
      updatedAt: now.toDate(),
      updatedById: input.updatedById,
      updatedByName: input.updatedByName,
    })
    const payload: SettingsDoc = {
      officeName: normalized.officeName,
      officeLatitude: normalized.officeLatitude,
      officeLongitude: normalized.officeLongitude,
      officeRadiusMeters: normalized.officeRadiusMeters,
      officePoints: normalized.officePoints.map((point) => ({
        id: point.id,
        name: point.name,
        latitude: point.latitude,
        longitude: point.longitude,
        radiusMeters: point.radiusMeters,
      })),
      updatedAt: now,
      updatedById: input.updatedById,
      updatedByName: input.updatedByName,
    }
    await setDoc(this.settingsRef, payload)
    return normalized
  }

  async getOfficeQr(dateKey: string): Promise<AttendanceOfficeQr | null> {
    const snapshot = await getDoc(doc(this.officeQrCollection, dateKey))
    if (!snapshot.exists()) return null
    return mapOfficeQr(snapshot.data() as OfficeQrDoc)
  }

  async saveOfficeQr(
    input: SaveAttendanceOfficeQrInput,
  ): Promise<AttendanceOfficeQr> {
    const now = Timestamp.now()
    const payload: OfficeQrDoc = {
      dateKey: input.dateKey,
      token: input.token,
      validFrom: Timestamp.fromDate(input.validFrom),
      validUntil: Timestamp.fromDate(input.validUntil),
      createdAt: now,
      createdById: input.createdById,
    }
    await setDoc(doc(this.officeQrCollection, input.dateKey), payload)
    return mapOfficeQr(payload)
  }

  async getByUserAndDate(
    userId: string,
    dateKey: string,
  ): Promise<Attendance | null> {
    const snapshot = await getDoc(
      doc(this.collectionRef, attendanceDocId(userId, dateKey)),
    )
    if (!snapshot.exists()) return null
    return mapAttendance(snapshot.id, snapshot.data() as AttendanceDoc)
  }

  async listByDate(dateKey: string): Promise<Attendance[]> {
    const snapshot = await getDocs(
      query(this.collectionRef, where('dateKey', '==', dateKey)),
    )
    return snapshot.docs
      .map((item) => mapAttendance(item.id, item.data() as AttendanceDoc))
      .filter((item): item is Attendance => item !== null)
      .sort((a, b) => a.createdAt.getTime() - b.createdAt.getTime())
  }

  async create(input: CreateAttendanceInput): Promise<Attendance> {
    const id = attendanceDocId(input.userId, input.dateKey)
    const ref = doc(this.collectionRef, id)
    const existing = await getDoc(ref)
    if (existing.exists()) {
      throw new ValidationError('Ya marcaste asistencia hoy')
    }

    const now = Timestamp.now()
    const payload: AttendanceDoc = {
      userId: input.userId,
      userName: input.userName,
      dateKey: input.dateKey,
      origin: input.origin,
      areaId: input.areaId,
      areaName: input.areaName,
      latitude: input.location.latitude,
      longitude: input.location.longitude,
      officeValidated: input.officeValidated,
      createdAt: now,
    }
    if (typeof input.location.accuracyMeters === 'number') {
      payload.locationAccuracy = input.location.accuracyMeters
    }
    if (typeof input.distanceToOfficeMeters === 'number') {
      payload.distanceToOfficeMeters = input.distanceToOfficeMeters
    }
    if (input.officeQrToken) {
      payload.officeQrToken = input.officeQrToken
    }
    if (input.permissionNote) {
      payload.permissionNote = input.permissionNote
    }
    if (input.markedById && input.markedByName) {
      payload.markedById = input.markedById
      payload.markedByName = input.markedByName
    }
    if (input.environmentPhotoUrl && input.environmentPhotoPath) {
      payload.environmentPhotoUrl = input.environmentPhotoUrl
      payload.environmentPhotoPath = input.environmentPhotoPath
    }

    try {
      await setDoc(ref, payload)
    } catch (error) {
      mapFirebaseWriteError(error, 'No se pudo marcar la asistencia')
    }
    return mapAttendance(id, payload) as Attendance
  }
}

File: consorcio/lib/infrastructure/firestore/firebase_attendance_repository.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_storage/firebase_storage.dart';

import '../../domain/entities/attendance.dart';
import '../../domain/entities/attendance_settings.dart';
import '../../domain/errors/domain_exception.dart';
import '../../domain/repositories/attendance_repository.dart';
import '../../domain/repositories/folder_image_repository.dart';
import '../../domain/value_objects/geo_location.dart';

class FirebaseAttendanceRepository implements AttendanceRepository {
  FirebaseAttendanceRepository({
    FirebaseFirestore? firestore,
    FirebaseStorage? storage,
  })  : _firestore = firestore ?? FirebaseFirestore.instance,
        _storage = storage ?? FirebaseStorage.instance;

  final FirebaseFirestore _firestore;
  final FirebaseStorage _storage;

  CollectionReference<Map<String, dynamic>> get _attendances =>
      _firestore.collection('attendances');

  DocumentReference<Map<String, dynamic>> get _settings =>
      _firestore.collection('settings').doc('attendance');

  @override
  Future<AttendanceSettings> getSettings() async {
    final snapshot = await _settings.get();
    if (!snapshot.exists || snapshot.data() == null) {
      return AttendanceSettings.defaults;
    }
    final data = snapshot.data()!;
    final rawPoints = data['officePoints'];
    final officePoints = rawPoints is List
        ? rawPoints
            .whereType<Map<String, dynamic>>()
            .map(
              (point) => AttendanceOfficePoint(
                id: point['id'] as String? ?? 'legacy',
                name: point['name'] as String? ??
                    AttendanceSettings.defaults.officeName,
                latitude: (point['latitude'] as num?)?.toDouble() ??
                    AttendanceSettings.defaults.officeLatitude,
                longitude: (point['longitude'] as num?)?.toDouble() ??
                    AttendanceSettings.defaults.officeLongitude,
                radiusMeters: AttendanceSettings.normalizeRadius(
                  (point['radiusMeters'] as num?)?.toInt() ??
                      AttendanceSettings.defaults.officeRadiusMeters,
                ),
              ),
            )
            .toList()
        : <AttendanceOfficePoint>[];

    final officeName =
        data['officeName'] as String? ?? AttendanceSettings.defaults.officeName;
    final officeLatitude = (data['officeLatitude'] as num?)?.toDouble() ??
        AttendanceSettings.defaults.officeLatitude;
    final officeLongitude = (data['officeLongitude'] as num?)?.toDouble() ??
        AttendanceSettings.defaults.officeLongitude;
    final officeRadiusMeters = AttendanceSettings.normalizeRadius(
      (data['officeRadiusMeters'] as num?)?.toInt() ??
          AttendanceSettings.defaults.officeRadiusMeters,
    );

    return AttendanceSettings(
      officePoints: officePoints.isNotEmpty
          ? officePoints
          : [
              AttendanceOfficePoint(
                id: 'legacy',
                name: officeName,
                latitude: officeLatitude,
                longitude: officeLongitude,
                radiusMeters: officeRadiusMeters,
              ),
            ],
      officeName: officeName,
      officeLatitude: officeLatitude,
      officeLongitude: officeLongitude,
      officeRadiusMeters: officeRadiusMeters,
    );
  }

  @override
  Future<Attendance?> getByUserAndDate(String userId, String dateKey) async {
    final snapshot =
        await _attendances.doc(attendanceDocId(userId, dateKey)).get();
    if (!snapshot.exists || snapshot.data() == null) return null;
    return _map(snapshot.id, snapshot.data()!);
  }

  @override
  Future<Attendance> create({
    required String userId,
    required String userName,
    required String dateKey,
    required AttendanceOrigin origin,
    required String areaId,
    required String areaName,
    required GeoLocation location,
    required bool officeValidated,
    int? distanceToOfficeMeters,
    String permissionNote = '',
    ImageFilePayload? environmentPhoto,
  }) async {
    final id = attendanceDocId(userId, dateKey);
    final ref = _attendances.doc(id);
    final existing = await ref.get();
    if (existing.exists) {
      throw DomainException('Ya marcaste asistencia hoy');
    }

    String? photoUrl;
    String? storagePath;
    if (environmentPhoto != null) {
      storagePath = 'attendances/$userId/$dateKey/entorno.jpg';
      final storageRef = _storage.ref(storagePath);
      try {
        await storageRef.putData(
          environmentPhoto.bytes,
          SettableMetadata(
            contentType: environmentPhoto.contentType.isEmpty
                ? 'image/jpeg'
                : environmentPhoto.contentType,
          ),
        );
        photoUrl = await storageRef.getDownloadURL();
      } catch (_) {
        throw DomainException(
          'No se pudo subir la foto. Revisa tu conexión.',
        );
      }
    }

    final now = Timestamp.now();
    final payload = <String, dynamic>{
      'userId': userId,
      'userName': userName,
      'dateKey': dateKey,
      'origin': origin.firestoreValue,
      'areaId': areaId,
      'areaName': areaName,
      'latitude': location.latitude,
      'longitude': location.longitude,
      'officeValidated': officeValidated,
      'createdAt': now,
    };
    if (photoUrl != null && storagePath != null) {
      payload['environmentPhotoUrl'] = photoUrl;
      payload['environmentPhotoPath'] = storagePath;
    }
    if (location.accuracyMeters != null) {
      payload['locationAccuracy'] = location.accuracyMeters;
    }
    if (distanceToOfficeMeters != null) {
      payload['distanceToOfficeMeters'] = distanceToOfficeMeters;
    }
    final trimmedNote = permissionNote.trim();
    if (trimmedNote.isNotEmpty) {
      payload['permissionNote'] =
          trimmedNote.length > 200 ? trimmedNote.substring(0, 200) : trimmedNote;
    }
    try {
      await ref.set(payload);
    } on FirebaseException catch (error) {
      if (error.code == 'permission-denied') {
        throw DomainException(
          'No se pudo registrar. Revisa el GPS, el radio de oficina o si ya hay una marca hoy.',
        );
      }
      throw DomainException('No se pudo marcar la asistencia');
    }
    return _map(id, payload);
  }

  Attendance _map(String id, Map<String, dynamic> data) {
    return Attendance(
      id: id,
      userId: data['userId'] as String? ?? '',
      userName: data['userName'] as String? ?? '',
      dateKey: data['dateKey'] as String? ?? '',
      origin: AttendanceOrigin.fromString(data['origin'] as String? ?? 'OFICINA'),
      areaId: data['areaId'] as String? ?? '',
      areaName: data['areaName'] as String? ?? '',
      location: GeoLocation(
        latitude: (data['latitude'] as num?)?.toDouble() ?? 0,
        longitude: (data['longitude'] as num?)?.toDouble() ?? 0,
        accuracyMeters: (data['locationAccuracy'] as num?)?.toDouble(),
        capturedAt: (data['createdAt'] as Timestamp?)?.toDate(),
      ),
      officeValidated: data['officeValidated'] == true,
      distanceToOfficeMeters: (data['distanceToOfficeMeters'] as num?)?.toInt(),
      createdAt: (data['createdAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
      environmentPhotoUrl: data['environmentPhotoUrl'] as String?,
      permissionNote: data['permissionNote'] as String?,
    );
  }
}
